ext/strings.go

```diff
@@ -485,6 +521,509 @@ func join(strs []string) (string, error) {
+	}
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		strException := false
+		if arg.Type() == types.StringType {
+			argStr := arg.Value().(string)
+			if argStr == "NaN" || argStr == "Infinity" || argStr == "-Infinity" {
+				strException = true
+			}
+		}
+		if arg.Type() != types.DoubleType && !strException {
+			return "", fmt.Errorf("fixed-point clause can only be used on doubles, was given %s", arg.Type().TypeName())
+		}
+		argFloatVal := arg.ConvertToType(types.DoubleType)
+		argFloat, ok := argFloatVal.Value().(float64)
+		if !ok {
+			return "", fmt.Errorf("could not convert \"%s\" to float64", argFloatVal.Value())
+		}
+		fmtStr := fmt.Sprintf("%%.%df", *precision)
+		matchedLocale := matchLanguage(locale)
+		return message.NewPrinter(matchedLocale).Sprintf(fmtStr, argFloat), nil
+	}, nil
+}
+
+func formatScientific(lastStrIndex int, formatStr string, useUpper bool, precision *int) (int, clauseImpl, error) {
+	if precision == nil {
+		precision = new(int)
+		*precision = defaultPrecision
+	}
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		strException := false
+		if arg.Type() == types.StringType {
+			argStr := arg.Value().(string)
+			if argStr == "NaN" || argStr == "Infinity" || argStr == "-Infinity" {
+				strException = true
+			}
+		}
+		if arg.Type() != types.DoubleType && !strException {
+			return "", fmt.Errorf("scientific clause can only be used on doubles, was given %s", arg.Type().TypeName())
+		}
+		argFloatVal := arg.ConvertToType(types.DoubleType)
+		argFloat, ok := argFloatVal.Value().(float64)
+		if !ok {
+			return "", fmt.Errorf("could not convert \"%s\" to float64", argFloatVal.Value())
+		}
+		var fmtStr string
+		if useUpper {
+			fmtStr = fmt.Sprintf("%%%dE", *precision)
+		} else {
+			fmtStr = fmt.Sprintf("%%%de", *precision)
+		}
+		matchedLocale := matchLanguage(locale)
+		return message.NewPrinter(matchedLocale).Sprintf(fmtStr, argFloat), nil
+	}, nil
+}
+
+func formatBinary(lastStrIndex int, formatStr string) (int, clauseImpl, error) {
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argIntVal := arg.ConvertToType(types.IntType)
+		argInt, ok := argIntVal.Value().(int64)
+		if !ok {
+			return "", fmt.Errorf("could not convert \"%s\" to int64", argIntVal.Value())
+		}
+		// locale is intentionally unused as integers formatted as binary
+		// strings are locale-independent
+		return fmt.Sprintf("%b", argInt), nil
+	}, nil
+}
+
+func formatHex(lastStrIndex int, formatStr string, useUpper bool) (int, clauseImpl, error) {
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		// special case to print out strings/byte arrays
+		if arg.Type() == types.StringType || arg.Type() == types.BytesType {
+			var buffer strings.Builder
+			fmtStr := "%x"
+			if useUpper {
+				fmtStr = "%X"
+			}
+			if arg.Type() == types.BytesType {
+				for _, b := range arg.Value().([]byte) {
```

Comment:
Why iterate over bytes and strings in this function instead of passing []byte directly into fmt.SPrintf("%x", ...)?

Reply:
Whoops, you're right. I've refactored both cases.